### PR TITLE
Changes for week of july 16 and 22

### LINKED ISSUE #141: Resource/Version model alignment changes
Proposal:
- add `versionid` to Resources and Versions - this will be the ID of the Version being shown. In the Resource's case this will always be the default Version's ID
- repurpose `id` on Version to be the ID of the Resource
- the `versions` map Key/ID is now the Version's `versionid` not `id`
- show all Version properties (including `isdefault`) in the Resource. This will allow a Resource's URL and a Version's URL to be interchangeable at least w.r.t. the Version's attributes. The remaining attributes in the Resource entity will then be more like administrative attributes
- Resource admin attributes become:
  - `defaultversionid`, `defaultversionurl`, `defaultversionsticky`, `versionscount`, `versionsurl`, `versions` (when inlining)
  - notice the rename of `stickydefaultversion` - to have it start with `defaultversion`, so `defaultversionsticky` now
  - notice that we don't have `versionurl` like we once did during our brainstorming on this, it's `defaultversionurl`

```
{
  "id": "STRING",                          # The Resource id
  "self": "URL",                           # Resource URL, not Version

  # These are inherited from the default Version (excluded on export)
  "versionid": "STRING",                   # Same a defaultversionid
  "name": "STRING", ?
  "epoch": UINTEGER,
  "isdefault": true,
  "description": "STRING", ?
  "documentation": "URL", ?
  "labels": { "STRING": "STRING" * }, ?
  "origin": "URI", ?
  "createdat": "TIME",
  "modifiedat": "TIME",
  "contenttype": "STRING, ?

  "RESOURCEurl": "URL", ?                  # If not local
  "RESOURCE": ... Resource document ..., ? # If local & inlined & JSON
  "RESOURCEbase64": "STRING", ?            # If local & inlined & ~JSON

  # These are Resource-only attributes
  "xref": "URL", ?                         # Ptr to other Resource

  "defaultversionsticky": BOOLEAN, ?
  "defaultversionid": "STRING",            # Same as versionid above
  "defaultversionurl": "URL",

  "versionsurl": "URL",
  "versionscount": UINTEGER,
  "versions": {                            # Only if inlined/nested
    "ID": {                                # The Version's versionid
      "id": "STRING",                      # The Resource id
      "self": "URL",                       # Version URL
      "versionid": "STRING",
      "name": "STRING", ?
      "epoch": UINTEGER,
      "isdefault": BOOLEAN, ?
      "description": "STRING", ?
      "documentation": "URL", ?
      "labels": { "STRING": "STRING" * }, ?
      "origin": "URI", ?
      "createdat": "TIME",
      "modifiedat": "TIME",
      "contenttype": "STRING", ?

      "RESOURCEurl": "URL", ?                  # If not local
      "RESOURCE": ... Resource document ..., ? # If inlined & JSON
      "RESOURCEbase64": "STRING" ?             # If inlined & ~JSON
    } *
  }
}
```

### PR DESCRIPTION
**Pull request by duglin**
Changes for week of July 16
- #141 : make the URL for a Resource and a Version be almost interchangeable
  - mainly add "versionid" to Versions
  - make Resource specific attribute all start with "versions..." or "defaultversion..."

Fixed #141

```
{
  "id": "STRING",                          # The Resource id
  "self": "URL",                           # Resource URL, not Version

  # These are inherited from the default Version (excluded on export)
  "versionid": "STRING",                   # Same a defaultversionid
  "name": "STRING", ?
  "epoch": UINTEGER,
  "isdefault": true,
  "description": "STRING", ?
  "documentation": "URL", ?
  "labels": { "STRING": "STRING" * }, ?
  "origin": "URI", ?
  "createdat": "TIME",
  "modifiedat": "TIME",
  "contenttype": "STRING, ?

  "RESOURCEurl": "URL", ?                  # If not local
  "RESOURCE": ... Resource document ..., ? # If local & inlined & JSON
  "RESOURCEbase64": "STRING", ?            # If local & inlined & ~JSON

  # These are Resource-only attributes
  "xref": "URL", ?                         # Ptr to other Resource

  "defaultversionsticky": BOOLEAN, ?
  "defaultversionid": "STRING",            # Same as versionid above
  "defaultversionurl": "URL",

  "versionsurl": "URL",
  "versionscount": UINTEGER,
  "versions": {                            # Only if inlined/nested
    "ID": {                                # The Version's versionid
      "id": "STRING",                      # The Resource id
      "self": "URL",                       # Version URL
      "versionid": "STRING",
      "name": "STRING", ?
      "epoch": UINTEGER,
      "isdefault": BOOLEAN, ?
      "description": "STRING", ?
      "documentation": "URL", ?
      "labels": { "STRING": "STRING" * }, ?
      "origin": "URI", ?
      "createdat": "TIME",
      "modifiedat": "TIME",
      "contenttype": "STRING", ?

      "RESOURCEurl": "URL", ?                  # If not local
      "RESOURCE": ... Resource document ..., ? # If inlined & JSON
      "RESOURCEbase64": "STRING" ?             # If inlined & ~JSON
    } *
  }
}
```

### CODE REVIEW

File: core/spec.md [hunk 1]
```diff
@@ -3692,8 +3688,8 @@ Where:
   of the 3 `RESOURCE` attributes MUST follow these rules:
```
Comment (deissnerk) [inline core/spec.md:3688]: The dot at the end of the previous line needs to be removed.

File: core/spec.md [hunk 2]
```diff
@@ -4060,27 +4072,66 @@ When serialized as a JSON object, the Version entity adheres to this form:
...
+- Constraints:
+  - MUST be a non-empty string consisting of [RFC3986 `unreserved`
+    characters](https://datatracker.ietf.org/doc/html/rfc3986#section-2.3)
+    (ALPHA / DIGIT / "-" / "." / "_" / "~").
+  - MUST NOT use a value of `null` or `request` due to these being reserved
+    for use by the `setdefaultversionid` operation.
+  - MUST be case insensitive unique within the scope of the owning Resource.
+  - This attribute MUST be treated as case sensitive for look-up purposes.
+    This means that an HTTP URL request to an entity with the wrong case for its
+    `versionid` MUST treated as "not found".
```
Comment (deissnerk) [inline core/spec.md:4117]:
```suggestion
    `versionid` MUST be treated as "not found".
```

File: core/spec.md [hunk 3]
```diff
@@ -4103,28 +4154,102 @@ and the following Version specific attributes:
...
+- Description: This attribute is a serialization of the corresponding
+  Resource document's contents. If the document bytes "as is" allow for them to
+  appear as the value of this attribute, then this attribute MUST be used if
+  the request asked for the document to be inlined in the response.
+- Constraints
+  - MUST NOT be present when the Resource's Registry metadata are being
+    serialized as HTTP headers.
+  - If the Resource's document is to be serialized and is not empty,
+    then either `RESOURCE` or `RESOURCEbase64` MUST be present.
+  - MUST only be used if the Resource document (bytes) are in the same
```
Comment (deissnerk) [inline core/spec.md:4206]: Not sure about this one
```suggestion
  - MUST only be used if the Resource document (bytes) is in the same
```

File: core/spec.md [hunk 4]
```diff
@@ -4103,28 +4154,102 @@ and the following Version specific attributes:
...
+    format as the Registry Resource entity.
+  - MUST NOT be present if `RESOURCEbase64` is also present.
+  - MUST NOT be present if the Resource's `hasdocument` model attribute is
+    set to `false.
+
+##### `RESOURCEbase64` Attribute
+- Type: String
+- Description: This attribute is a base64 serialization of the corresponding
+  Resource document's contents. If the Resource document (which are stored as
+  an array of bytes) are not conformant with the format being used to serialize
```
Comment (deissnerk) [inline core/spec.md:4216]:
```suggestion
  Resource document's contents. If the Resource document (which is stored as
  an array of bytes) is not conformant with the format being used to serialize
```

File: core/spec.md [hunk 5]
```diff
@@ -4103,28 +4154,102 @@ and the following Version specific attributes:
...
+  with the Resource object (i.e. as a JSON value), then this attribute MUST be
+  used in instead of the `RESOURCE` attribute.
+- Constraints:
+  - MUST NOT be present when the Resource's Registry metadata are being
```
Comment (deissnerk) [inline core/spec.md:4220]: Not sure what's correct.
```suggestion
  - MUST NOT be present when the Resource's Registry metadata is being
```

File: core/spec.md [hunk 6]
```diff
@@ -4103,28 +4154,102 @@ and the following Version specific attributes:
...
-
-Servers MAY have their own algorithm for creation of new Version `id` values,
-but the default algorithm is as follows:
-- `id` MUST be a string serialization of a monotonically increasing (by `1`)
-  unsigned integer starting with `1` and is scoped to the owning Resource.
-- Each time a new `id` is needed, if an existing Version already has that
-  `id` then the server MUST generate the next `id` value and try again.
+If a server does not support client-side specification of the `versionid` of a
+new Version (see the `setversionid` attribute in the [Registry
+Model](#registry-model), or if a client choose to not specify the `versionid`,
```
Comment (deissnerk) [inline core/spec.md:4233]:
```suggestion
Model](#registry-model), or if a client chooses to not specify the `versionid`,
```

File: core/spec.md [hunk 7]
```diff
@@ -4103,28 +4154,102 @@ and the following Version specific attributes:
...
+then the server MUST assign new Version an `versionid` that is unique within
+the scope of its owning Resource.
+
+Servers MAY have their own algorithm for creation of new Version `versionid`
```
Comment (deissnerk) [inline core/spec.md:4237]:
```suggestion
Servers MAY have their own algorithm for the creation of new Version `versionid`
```

File: core/spec.md [hunk 8]
```diff
@@ -4103,28 +4154,102 @@ and the following Version specific attributes:
...
+values, but the default algorithm is as follows:
+- `versionid` MUST be a string serialization of a monotonically increasing
+  (by `1`) unsigned integer starting with `1` and is scoped to the owning
+  Resource.
+- Each time a new `versionid` is needed, if an existing Version already has
+  that `versionid` then the server MUST generate the next `versionid` value
+  and try again.
 - The search for the next value does not restart with `1` each time, it MUST
   continue from the previously generated value.
```
Comment (deissnerk) [inline core/spec.md:4246]: This is even true if the version with the previously generated value got deleted again, correct? By default, the "re-use" of version ids should be avoided.

File: core/spec.md [hunk 9]
```diff
@@ -4564,8 +4698,9 @@ This means that any Registry Collection's attributes MUST be modified
 to match the resulting subset. In particular:
 - If the collection is inlined it MUST only include entities that match the
   filter expression(s).
-- The collection `url` attribute MUST include the filter expression(s) in its
-  query parameters.
+- The collection `url` attribute MUST include the appropriate filter
+  expression(s) in its query parameters such that a an HTTP `GET` to that URL
```
Comment (deissnerk) [inline core/spec.md:4702]:
```suggestion
  expression(s) in its query parameters such that an HTTP `GET` to that URL
```

**Review by deissnerk — COMMENTED** [review]
Mostly typos. Sometimes I wasn't sure.

File: core/spec.md [hunk 10]
```diff
@@ -230,8 +237,8 @@ For easy reference, the JSON serialization of a Registry adheres to this form:
 Note: the cardinality/optionality of a Resource's attributes are specified
 to indicate which are to be present when the Resource's attributes are
 included in its serialization. For example, when `?export` is used, the
-Resource's default attributes will not be present at all, which means listing
-`epoch` as mandatory in the above serialize is technically incorrect in
+Resource's default Version attributes will not be present at all, which means
+listing `epoch` as mandatory in the above serialize is technically incorrect in
```
[the remark this answers is not in the document]
Comment (duglin) [reply core/spec.md:241]: done

File: core/spec.md [hunk 11]
```diff
@@ -334,8 +341,8 @@ Unless otherwise noted, all attributes and extensions MUST be mutable and MUST
 be one of the following data types:
 - `any` - an attribute of this type is one whose type is not known in advance
    and MUST be one of the concrete types listed here.
-- `array` - an ordered set of values whose values are all of the same data
-   type - one of the types listed here.
+- `array` - an ordered set of values that are all of the same data type - one
```
[the remark this answers is not in the document]
Comment (duglin) [reply core/spec.md:344]: done

File: core/spec.md [hunk 12]
```diff
@@ -807,16 +800,18 @@ Where:
...
   be a read-only attribute that MUST be silently ignored by a server during
   an write operation.
 - The `COLLECTIONs` attribute is a map and MUST contain the entities of the
   collection (after any necessary [filtering](#filtering)), and MUST use
-  each entity's `id` as the key for that map entry.
-- The `id` for each entity in the collection MUST be unique within the scope
-  of the collection.
+  the Group's `id`, Resource's `id` or Version's `versionid` as the key for
+  that map entry.
+- The key each entity in the collection MUST be unique within the scope of the
```
[the remark this answers is not in the document]
Comment (duglin) [reply core/spec.md:813]: done

File: core/spec.md [hunk 13]
```diff
@@ -956,6 +954,8 @@ this cases the `name` attributes have different values. The paragraph above
 mandates that in these potentially ambiguous cases the entity in the
 `versions` collection is to be used and the top-level attributes are to be
 ignored - for the purposes of updating the "default" Version's attributes.
+So, in this case the `name` of the default(`v1`) Verison will be
```
[the remark this answers is not in the document]
Comment (duglin) [reply core/spec.md:957]: done. But I'm wondering why the spell-checker didn't notice this... I'll look into that

File: core/spec.md [hunk 14]
```diff
@@ -1186,15 +1192,17 @@ Where:
...
-- If an `id` value is present in the object, then it MUST match its
-  corresponding key value.
+  more entries where the key of each entity is each entity's unique
+  identifier - which is the `id` for Groups and Resources, and the `versionid`
+  for Versions.
 - If an `epoch` value is specified for an entity then the server MUST check
   to ensure that the value matches the entity's current `epoch` value and if it
   differs then an error MUST be generated.
+- If the entity's unique identifier is present in the object, then it MUST
+  MUST match its corresponding key value. And for Versions, if `id` is present
```
[the remark this answers is not in the document]
Comment (duglin) [reply core/spec.md:1202]: done

File: core/spec.md [hunk 15]
```diff
@@ -1285,13 +1296,15 @@ and the following Registry specific attributes:
...
-**`GROUPs` collections**
+##### `GROUPs` Collections
 - Type: Set of [Registry Collections](#registry-collections)
 - Description: A list of Registry collections that contain the set of Groups
   supported by the Registry.
 - Constraints:
   - REQUIRED in responses, MAY be present in requests.
   - REQUIRED in document view.
+  - If present, it MUST include all nested Group Collection types in the
+    Registry, even if the some of the collections are empty.
```
[the remark this answers is not in the document]
Comment (duglin) [reply core/spec.md:1307]: done

File: core/spec.md [hunk 16]
```diff
@@ -1587,11 +1598,11 @@ with respect to how models are defined or updated:
 - The `name` and `type` aspects of attributes MUST NOT be changed.
 
 To indicate a change to a specification defined attribute, the attribute MUST
-be defined as part of the Registry model. Any specification defined attributes
-that are not explicitly defined in the model of a Registry MUST be implicitly
-present and be as defined by this specification. In other words, the
+be included  as part of the Registry model update request. Any specification
+defined attributes that are not explicitly defined in the request MUST be
+implicitly defined and be as defined by this specification. In other words, the
```
[the remark this answers is not in the document]
Comment (duglin) [reply core/spec.md:1603]: I took a try at cleaning it up - let me know

File: core/spec.md [hunk 17]
```diff
@@ -1815,10 +1828,13 @@ The following describes the attributes of Registry model:
 
 - `attributes."STRING".clientrequired`
   - Indicates whether this attribute is a REQUIRED field for a client when
-    creating or update an entity. During creation, or update, of an entity if
-    this attribute is not specified then an error MUST be generated. When not
-    specified the default value is `false`. When the attribute name is `*`
-    then `clientrequired` MUST NOT be set to `true`.
+    creating or update an entity. When not specified the default value is
```
[the remark this answers is not in the document]
Comment (duglin) [reply core/spec.md:1831]: done

File: core/spec.md [hunk 18]
```diff
@@ -1815,10 +1828,13 @@ The following describes the attributes of Registry model:
...
+    `false`. When the attribute name is `*` then `clientrequired` MUST NOT be
+    set to `true`.
+
+    During creation, or update, of an entity if this attribute is not
```
[the remark this answers is not in the document]
Comment (duglin) [reply core/spec.md:1835]: not sure - removed them :-)

File: core/spec.md [hunk 19]
```diff
@@ -2018,7 +2038,8 @@ The following describes the attributes of Registry model:
 - `groups.resources.readonly`
   - Indicates if Resources of this type are updateable by a client. A value of
     `false` means that the server MUST support write operations on to Resources
```
[the remark this answers is not in the document]
Comment (duglin) [reply core/spec.md:2040]: removed "to"

File: core/spec.md [hunk 20]
```diff
@@ -2032,14 +2053,14 @@ The following describes the attributes of Registry model:
...
+        can be serialized as a "string", possibly with some escaping.<br>
 
     For some well-known `contenttype` values (e.g. `application/json`) the
     first case can be easily determined by the server. However, for custom
-    values the server will need to be explicitly told to when to consider it
-    to be of the correct type. The `typemap` attribute allows for this by
-    defining a mapping of `contenttype` values to well-known xRegistry format
-    types.
+    `contenttype` values the server will need to be explicitly told how to
+    interpret its value (e.g. to know if it a string or JSON).
```
[the remark this answers is not in the document]
Comment (duglin) [reply core/spec.md:2061]: done

File: core/spec.md [hunk 21]
```diff
@@ -2579,28 +2602,31 @@ The serialization of a Group entity adheres to this form:
...
+##### `RESOURCEs` Collections
 - Type: Set of [Registry Collections](#registry-collections).
 - Description: A list of Registry collections that contain the set of
   Resources supported by the Group.
 - Constraints:
   - REQUIRED in responses, MAY be present in requests.
   - REQUIRED in document view.
   - If present in a response, it MUST include all nested Resource Collection
-    types of the owning Group.
+    types of the owning Group, even if the some of the collections are empty.
```
[the remark this answers is not in the document]
Comment (duglin) [reply core/spec.md:2629]: done

File: core/spec.md [hunk 22]
```diff
@@ -2898,25 +2924,13 @@ the Resource. The Resource entity can be thought of as an alias for the
...
-- `id` MUST be for the Resource itself and not the `id` from the "default"
-  Version. The `id` of each Version MUST be unique within the scope of the
-  Resource, but to ensure the Resource itself has a consistent `id` value
-  it is distinct from any Version's `id`.
-
-  Additionally, Resource `id` values are often human readable descriptions
-  of the Resource (e.g. `mySchema`), while Version `id` values are meant to be
-  version string values (e.g. `v1.0`).
-
+However, there is one exceptions:
```
[the remark this answers is not in the document]
Comment (duglin) [reply core/spec.md:2927]: done

File: core/spec.md (hunk 1 above)
Comment (duglin) [reply core/spec.md:3688]: done

File: core/spec.md (hunk 2 above)
Comment (duglin) [reply core/spec.md:4117]: done

File: core/spec.md (hunk 3 above)
Comment (duglin) [reply core/spec.md:4206]: let's do it!

File: core/spec.md (hunk 4 above)
Comment (duglin) [reply core/spec.md:4216]: done

File: core/spec.md (hunk 5 above)
Comment (duglin) [reply core/spec.md:4220]: I think you're right

File: core/spec.md (hunk 6 above)
Comment (duglin) [reply core/spec.md:4233]: done

File: core/spec.md (hunk 7 above)
Comment (duglin) [reply core/spec.md:4237]: done

File: core/spec.md (hunk 8 above)
Comment (duglin) [reply core/spec.md:4246]: yes. I added the word "highest" see if that helps clarify it

File: core/spec.md (hunk 9 above)
Comment (duglin) [reply core/spec.md:4702]: done

**Comment by duglin** [comment]
@deissnerk thanks for the very thorough review - I picked up all of your suggestions

**Comment by duglin** [comment]
Approved on the 7/25 call.
there are more changes needed to the other spec but I'll get that in a follow-on PR.

**Pull request merged**